File: APIdoc/java/apidoc/src/main/java/io/point/apidoc/tutorials/ListFoldersCheckInCheckOut.java
package io.point.apidoc.tutorials;

import io.point.apidoc.APIDoc;
import org.codehaus.jackson.JsonNode;
import org.codehaus.jackson.node.ArrayNode;

import java.util.Iterator;

/**
 * User: jconley
 * Date: 5/6/2014
 */
public class ListFoldersCheckInCheckOut {

    public static void main(String[] args) throws Exception{
        if(args.length != 3){
            System.err.println("Valid arguments are email password apiKey");
            System.exit(-1);
        }

        String email = args[0];
        String password = args[1];
        String apiKey = args[2];

        String sessionKey = APIDoc.authenticate(email, password, apiKey);

        ArrayNode accessRulesArray = APIDoc.listAccessRules(sessionKey);
        JsonNode firstRule = accessRulesArray.get(0);

        ArrayNode folderArray = APIDoc.listFolders(sessionKey, firstRule.get(1).getTextValue());
        Iterator<JsonNode> rootObjects = folderArray.getElements();
        while(rootObjects.hasNext()) {
            ArrayNode file = (ArrayNode)rootObjects.next();

            //if it's a director, create a link for it and exit
            if(file.get(2).getTextValue().equals("FILE")){
                APIDoc.checkout(sessionKey, file.get(16).asText(), file.get(0).asText(), file.get(1).asText());

                String downloadUrl = APIDoc.fileDownload(sessionKey, file.get(16).asText(), file.get(0).asText(), file.get(1).asText());

                //follow link and download file

                //edit file
                String localPath = "";

                APIDoc.fileUpload(sessionKey, file.get(16).asText(), file.get(0).asText(), file.get(1).asText(), localPath);

                APIDoc.checkin(sessionKey, file.get(16).asText(), file.get(0).asText(), file.get(1).asText());

                break;
            }
        }
    }
}

File: APIdoc/scala/apidoc/src/main/scala/io/point/apidoc/tutorials/ListFoldersCheckInCheckOut.scala
package io.point.apidoc.tutorials

import io.point.apidoc.APIDoc
import dispatch._, Defaults._
import play.api.libs.json.JsArray

/**
  * User: jconley
  * Date: 5/7/2014
  */
object ListFoldersCheckInCheckOut {

   def main(args: Array[String]) {
     if(args.length != 3){
       System.err.println("Invalid args, must be email password apiKey")
       System.exit(-1)
     }

     val (email, password, apiKey) = (args(0), args(1), args(2))

     for {
       keyJson <- APIDoc.authenticate(email, password, apiKey)
       sessionKey = keyJson.getOrElse(throw new Exception("Session key not found"))
       accessRulesJson <- APIDoc.listAccessRules(sessionKey)
       accessRules = accessRulesJson.getOrElse(throw new Exception("Access rules not found"))
       foldersJson <- APIDoc.listFolders(sessionKey, accessRules(2).as[JsArray].value(1).as[String])
       folders = foldersJson.getOrElse(throw new Exception("Folders not found"))
     } yield {
        folders.value.foreach{ f =>
        val folder = f.as[JsArray]
          if(folder.value(2).as[String] == "FILE") {
            for {
              checkout <- APIDoc.fileCheckout(sessionKey, folder.value(16).as[String], folder.value(0).as[String], folder.value(1).as[String])
              download <- APIDoc.fileDownload(sessionKey, folder.value(16).as[String], folder.value(0).as[String], folder.value(1).as[String])
              //edit file
              localPath = ""
              upload <- APIDoc.fileUpload(sessionKey, folder.value(16).as[String], folder.value(0).as[String], folder.value(1).as[String], localPath)
              checkin <- APIDoc.fileCheckin(sessionKey, folder.value(16).as[String], folder.value(0).as[String], folder.value(1).as[String])
            } yield {
              checkin
            }
          }
       }
     }
   }
 }
